### DISCUSSION TITLE
How to get records which were failed while bulk copy insert

### DISCUSSION THREAD

**Discussion by VarunSaiTeja** [linq2db/linq2db#3081] [Q&A] [answered]
I am trying to perform a bulk insert.
I am doing a bulk copy of multiple rows. There might be a chance some record that will throw an exception due to certain reasons. For the example, I shared it's due to duplicate PK i.e Serial.

I want to get info of records that were failed to insert for further processing. And I want the remaining records to be inserted into the expected table.

```
AppDbService appDb = new AppDbService();

appDb.Insert(new StorageInfo() { Serial = "A", Alias = "Orginal" });

var stt = new List<StorageInfo>();
stt.Add(new StorageInfo() { Serial = "B", Alias = "B" });

//Trying to insert record with duplicate PK i.e Serial
stt.Add(new StorageInfo() { Serial = "A", Alias = "Duplicate" });

stt.Add(new StorageInfo() { Serial = "C", Alias = "C" });
appDb.BulkCopy(stt);

//I expect the storage table should have the records A, B, C. And the record info { Serial = "A", Alias = "Duplicate" } need to be shared as the response of BulkCopy to process that record separately after clearing errors.
```

Is there any way to get records info which was failed in insertion?

### Environment details
linq2db version: 3.4.0
Database Server: Sqlite
Database Provider: Microsoft.Data.Sqlite
Operating system: Windows 10
.NET version: 5

**Comment by MaceWindu** [comment, marked as answer]
Technically it should be possible with:
1. use `BulkCopyType.RowByRow` mode
2. don't use transaction or continue insert in same transaction
3. set [BulkCopyOptions.NotifyAfter](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Data/BulkCopyOptions.cs#L84) to `1`
4. set notification callback [BulkCopyOptions.RowsCopiedCallback](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Data/BulkCopyOptions.cs#L90) to get notification after each insert with number of inserted records so you know last inserted record

But I don't see how it will be easier than insert in loop.

**Reply by jods4** [reply]
if you're ok with provider-specific stuff, there is a more efficient way, sometimes.
Oracle for example has an extension where rows that fail in an INSERT..SELECT are logged into another table and the overall insert doesn't fail.

**Reply by MaceWindu** [reply]
well, @VarunSaiTeja  specified sqlite. I'm not really aware of any db with partial insert maybe except one you mentioned...

**Comment by VarunSaiTeja** [comment]
I also had requirement of using at SQL server.
Is there any trick for sql server?

**Reply by MaceWindu** [reply]
Nope, both multi-row insert and bulk copy doesn't allow partial insert.
[MAXERRORS option](https://docs.microsoft.com/en-us/sql/t-sql/statements/bulk-insert-transact-sql?redirectedfrom=MSDN&view=sql-server-ver15) also doesn't help as it doesn't work for constraint errors and not available from SqlClient anyway.
